### PR TITLE
Add Twilio proof of concept for SMS to Download

### PR DESCRIPTION
**Pull request by robhudson**

### CODE REVIEW

File: requirements/docs.txt [hunk 1]
```diff
@@ -321,12 +321,6 @@ tornado==6.1 \
     --hash=sha256:fa2ba70284fa42c2a5ecb35e322e68823288a4251f9ba9cc77be04ae15eada68 \
     --hash=sha256:fba85b6cd9c39be262fcd23865652920832b61583de2a2ca907dbd8e8a8c81e5
     # via livereload
-typing-extensions==4.0.1 \
```
Comment (pmac) [inline requirements/docs.txt:324]: wonder what's up with these changes to docs requirements.

Comment (robhudson) [reply requirements/docs.txt:324]: Not sure. I ran the `make compile-requirements` and the popped up. I can drop them from this PR though.

Comment (pmac) [reply requirements/docs.txt:324]: interesting. might be fine. ¯\_(ツ)_/¯

File: bedrock/firefox/tests/test_views.py [hunk 2]
```diff
@@ -678,3 +676,60 @@ def test_firefox_welcome_page1(self, render_mock):
...
+        patcher = patch("bedrock.firefox.views.TWILIO_CLIENT")
+        self.mock_create = patcher.start()
+        self.addCleanup(patcher.stop)
+
+    def _do_post(self, params):
+        return self.client.post(
+            self.url,
+            data=params,
+            follow=True,
+            HTTP_X_REQUESTED_WITH="XMLHttpRequest",
```
Comment (pmac) [inline bedrock/firefox/tests/test_views.py:693]: I don't think this is required. I don't see a use of `request.is_ajax()` anywhere. Not going to hurt anything though.

File: bedrock/firefox/tests/test_views.py [hunk 3]
```diff
@@ -678,3 +676,60 @@ def test_firefox_welcome_page1(self, render_mock):
...
+        self.mock_create.messages.create.assert_called_once_with(
+            to="+14155551212",
+            body="test",
+            messaging_service_sid=settings.TWILIO_MESSAGING_SERVICE_SID,
+        )
+
+    def test_twilio_exception(self):
+        self.mock_create.messages.create.side_effect = TwilioRestException(status=500, uri="/", msg="Some error")
+        resp = self._do_post({"phone_number": "415 555 1212"})
+        assert resp.status_code == 200
```
Comment (pmac) [inline bedrock/firefox/tests/test_views.py:732]: should the API respond with a code in the 400s when there's an error?

**Review by pmac — APPROVED** [review]
This looks great! I added a couple of questions, but this can land whenever we're ready IMO.

File: bedrock/firefox/tests/test_views.py (hunk 3 above)
Comment (stevejalim) [reply bedrock/firefox/tests/test_views.py:732]: Agreed - if we're mocking out the Twilio client, would still be nice to get a 400 as per https://www.twilio.com/docs/verify/api/v1/error-codes

File: bedrock/firefox/views.py [hunk 4]
```diff
@@ -23,6 +24,8 @@
 import basket
 import querystringsafe_base64
 from product_details.version_compare import Version
+from twilio.base.exceptions import TwilioRestException
```
Comment (stevejalim) [inline bedrock/firefox/views.py:27]: Is this used? I can't see TwilioRestException below, but surely Flake8 would grumble

File: bedrock/firefox/views.py [hunk 5]
```diff
@@ -176,6 +191,33 @@ def sign_attribution_codes(codes):
...
+def sms_send_to_device_ajax(request):
+    form = SMSSendToDeviceForm(request.POST)
+    if not form.is_valid():
+        return JsonResponse({"success": False, "errors": form.errors["phone_number"]})
+
+    to_number = form.cleaned_data["phone_number"]
+    # platform = form.cleaned_data["platform"]
+
+    # TODO: Update `sms_body` with actual content. May differ based on platform.
+    sms_body = "test"  # Limited to 1600 characters.
```
Comment (stevejalim) [inline bedrock/firefox/views.py:205]: Wow 1600 chars for the SMS limit? I guess they'll split them for us - but if they do, will it not break a URL, for instance?

File: bedrock/firefox/views.py [hunk 6]
```diff
@@ -176,6 +191,33 @@ def sign_attribution_codes(codes):
     return {"attribution_code": code.decode(), "attribution_sig": sig}
 
 
+@require_POST
+@csrf_exempt
```
Comment (stevejalim) [inline bedrock/firefox/views.py:195]: Do we need to make this CSRF exempt? If you want to add CSRF back in, we can get the CSRF token via some boilerplate JS - https://docs.djangoproject.com/en/4.0/ref/csrf/#ajax

File: bedrock/settings/base.py [hunk 7]
```diff
@@ -1018,6 +1018,11 @@ def get_app_name(hostname):
     },
 }
 
+# Twilio
+TWILIO_ACCOUNT_SID = config("TWILIO_ACCOUNT_SID", default="")
+TWILIO_AUTH_TOKEN = config("TWILIO_AUTH_TOKEN", default="")
+TWILIO_MESSAGING_SERVICE_SID = config("TWILIO_MESSAGING_SERVICE_SID", default="")
```
Comment (stevejalim) [inline bedrock/settings/base.py:1024]: I guess we'll need a corresponding addition to www-config, or are there keys there from previous eras?

**Review by stevejalim — APPROVED** [review]
LGTM! Nice one 🎸  r+wc and most of those are bikesheddy comments 😆

File: bedrock/firefox/forms.py [hunk 8]
```diff
@@ -15,3 +17,29 @@ class SendToDeviceWidgetForm(forms.Form):
...
     )
+
+
+class SMSSendToDeviceForm(forms.Form):
+    phone_number = forms.CharField(max_length=20, required=False)
+    platform = forms.ChoiceField(
+        choices=(
+            ("ios", "ios"),
+            ("android", "android"),
+            ("all", "all"),
```
[the remark this answers is not in the document]
Comment (robhudson) [reply bedrock/firefox/forms.py:28]: There's similar code on the email to download form and I believe these are set automatically based on the user's selection when navigating to the page. So from what I can tell the end user never sees it.

File: bedrock/firefox/tests/test_views.py (hunk 3 above)
Comment (robhudson) [reply bedrock/firefox/tests/test_views.py:732]: I thought so too but the email to device endpoint doesn't so I thought perhaps it was a front-end preference to send 200s but with a `success` boolean for reporting to the end user.

File: bedrock/firefox/views.py (hunk 4 above)
Comment (robhudson) [reply bedrock/firefox/views.py:27]: Yeah there's a `try / except` down below where it is used.

File: bedrock/settings/base.py (hunk 7 above)
Comment (robhudson) [reply bedrock/settings/base.py:1024]: No keys that I know of, so I'll need to update that. Thanks for the call out.

**Comment by robhudson** [comment]
Closing this PR. Opened a branch at https://github.com/mozilla/bedrock/tree/twilio-poc and will work with FE to complete this feature.

**Pull request closed without merging**